<a href="https://colab.research.google.com/github/aryann99/Machine_Learning_Models/blob/main/financial_market_news.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**FINANCIAL MARKET ANALYSIS**

**OBJECTIVE**

The objective of a financial market news analysis ML model is to extract insights and make predictions based on the information available in financial news articles. The primary goals of such a model can include:

Sentiment analysis: Determining the sentiment expressed in financial news articles, such as whether the article is positive, negative, or neutral. This can help investors and traders gauge market sentiment and make informed decisions.

Market impact assessment: Analyzing the potential impact of news events on financial markets. By identifying and understanding news articles that have a significant influence on market movements, investors can react swiftly and strategically.

**import library**

In [ ]:
import pandas as pd

In [ ]:
import numpy as np

**import data**

In [ ]:
df = pd.read_csv(r'https://github.com/YBIFoundation/Dataset/blob/main/Financial%20Market%20News.csv?raw=true', encoding="ISO-8859-1")

In [ ]:
df.head()

Date  Label                                             News 1  \
0  01-01-2010      0           McIlroy's men catch cold from Gudjonsson   
1  02-01-2010      0               Warning from history points to crash   
2  03-01-2010      0      Comment: Why Israel's peaceniks feel betrayed   
3  04-01-2010      1             £750,000-a-goal Weah aims parting shot   
4  05-01-2010      1  Leeds arrive in Turkey to the silence of the fans   

                                       News 2  \
0                       Obituary: Brian Walsh   
1              Investors flee to dollar haven   
2  Court deals blow to seizure of drug assets   
3            Newcastle pay for Fletcher years   
4     One woman's vision offers loan lifeline   

                                            News 3  \
0       Workplace blues leave employers in the red   
1                      Banks and tobacco in favour   
2                       An ideal target for spooks   
3  Brown sent to the stands for Scotland qualifier   
4          Working Lives: How world leaders worked   

                                        News 4  \
0                     Classical review: Rattle   
1                        Review: Llama Farmers   
2  World steps between two sides intent on war   
3         Tourists wary of breaking new ground   
4           Working Lives: Tricks of the trade   

                                              News 5  \
0                     Dance review: Merce Cunningham   
1                       War jitters lead to sell-off   
2                       What the region's papers say   
3              Canary Wharf climbs into the FTSE 100   
4  Working Lives: six-hour days, long lunches and...   

                                         News 6  \
0  Genetic tests to be used in setting premiums   
1                    Your not-so-secret history   
2           Comment: Fear and rage in Palestine   
3                           Review: Bill Bailey   
4                        Pop review: We Love UK   

                                          News 7  \
0                        Opera review: La Bohème   
1                  Review: The Northern Sinfonia   
2  Poverty and resentment fuels Palestinian fury   
3                              Review: Classical   
4               World music review: Marisa Monte   

                                        News 8  ...  \
0                   Pop review: Britney Spears  ...   
1                             Review: Hysteria  ...   
2  Republican feud fear as dissident is killed  ...   
3              Review: New Contemporaries 2000  ...   
4             Art review:  Hollingsworth/Heyer  ...   

                                News 16                              News 17  \
0                 Finland 0 - 0 England                   Healy a marked man   
1  Why Wenger will stick to his Gunners  Out of luck England hit rock bottom   
2             FTSE goes upwardly mobile             At this price?  BP Amoco   
3         More cash on way for counties      Cairns carries Kiwis to victory   
4          Duisenberg in double trouble           Pru to cut pension charges   

                                             News 18  \
0                     Happy birthday Harpers & Queen   
1                         Wilkinson out of his depth   
2                                            Go fish   
3  Year of Blanchflower's flourish when Spurs sto...   
4                            Art review: Paul Graham   

                                             News 19  \
0  Win unlimited access to the Raindance film fes...   
1                   Kinsella sparks Irish power play   
2         Bosnian Serb blows himself up to evade law   
3      New direct approach brings only pay-per-blues   
4               Shearer shot sparks Boro humiliation   

                                             News 20  \
0  Labour pledges £800m to bridge north-south divide   
1                    Brown banished as Scots rebound   
2                       Or

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101 entries, 0 to 4100
Data columns (total 27 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Date     4101 non-null   object
 1   Label    4101 non-null   int64 
 2   News 1   4101 non-null   object
 3   News 2   4101 non-null   object
 4   News 3   4101 non-null   object
 5   News 4   4101 non-null   object
 6   News 5   4101 non-null   object
 7   News 6   4101 non-null   object
 8   News 7   4101 non-null   object
 9   News 8   4101 non-null   object
 10  News 9   4101 non-null   object
 11  News 10  4101 non-null   object
 12  News 11  4101 non-null   object
 13  News 12  4101 non-null   object
 14  News 13  4101 non-null   object
 15  News 14  4101 non-null   object
 16  News 15  4101 non-null   object
 17  News 16  4101 non-null   object
 18  News 17  4101 non-null   object
 19  News 18  4101 non-null   object
 20  News 19  4101 non-null   object
 21  News 20  4101 non-null   object
 22  

In [ ]:
df.shape

(4101, 27)

In [ ]:
df.columns

Index(['Date', 'Label', 'News 1', 'News 2', 'News 3', 'News 4', 'News 5',
       'News 6', 'News 7', 'News 8', 'News 9', 'News 10', 'News 11', 'News 12',
       'News 13', 'News 14', 'News 15', 'News 16', 'News 17', 'News 18',
       'News 19', 'News 20', 'News 21', 'News 22', 'News 23', 'News 24',
       'News 25'],
      dtype='object')

In [ ]:
#feature selection
' '.join(str(x) for x in df.iloc[1,2:27])

"Warning from history points to crash Investors flee to dollar haven Banks and tobacco in favour Review: Llama Farmers War jitters lead to sell-off Your not-so-secret history Review: The Northern Sinfonia Review: Hysteria Review: The Guardsman Opera: The Marriage of Figaro Review: The Turk in Italy Deutsche spells out its plans for diversification Traders' panic sends oil prices skyward TV sport chief leaves home over romance Leader: Hi-tech twitch Why Wenger will stick to his Gunners Out of luck England hit rock bottom Wilkinson out of his depth Kinsella sparks Irish power play Brown banished as Scots rebound Battling Wales cling to lifeline Ehiogu close to sealing Boro move Man-to-man marking Match stats French referee at centre of storm is no stranger to controversy"

In [ ]:
df.index

RangeIndex(start=0, stop=4101, step=1)

In [ ]:
len(df.index)

4101

In [ ]:
news = []
for row in range(0, len(df.index)):
  news.append(' '.join(str(x) for x in df.iloc[row,2:27]))

In [ ]:
type(news)

list

In [ ]:
news[0]

"McIlroy's men catch cold from Gudjonsson Obituary: Brian Walsh Workplace blues leave employers in the red Classical review: Rattle Dance review: Merce Cunningham Genetic tests to be used in setting premiums Opera review: La Bohème Pop review: Britney Spears Theatre review: The Circle Wales face a fraught night Under-21  round-up Smith off to blot his copybook Finns taking the mickey Praise wasted as Brown studies injury options Ireland wary of minnows Finland 0 - 0 England Healy a marked man Happy birthday Harpers & Queen Win unlimited access to the Raindance film festival Labour pledges £800m to bridge north-south divide Wales: Lib-Lab pact firm despite resignation Donald Dewar Regenerating homes  regenerates well-being in people Win £100 worth of underwear TV guide: Random views"

In [ ]:
x = news

In [ ]:
type(x)

list

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
cv = CountVectorizer (lowercase =   True, ngram_range=(1,1))

**Define Target Variable (y) and Feature Variables (X)**

In [ ]:
x = cv.fit_transform(x)

In [ ]:
x.shape

(4101, 48527)

In [ ]:
y = df['Label']

In [ ]:
y.shape

(4101,)

**Train Test Split**

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size = 0.3, stratify=y, random_state=2529)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf = RandomForestClassifier(n_estimators=200)

In [ ]:
rf.fit(x_train, y_train)

RandomForestClassifier(n_estimators=200)

In [ ]:
y_pred = rf.predict(x_test)

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [ ]:
confusion_matrix(y_test, y_pred)

array([[134, 447],
       [170, 480]])

In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.44      0.23      0.30       581
           1       0.52      0.74      0.61       650

    accuracy                           0.50      1231
   macro avg       0.48      0.48      0.46      1231
weighted avg       0.48      0.50      0.46      1231



**Explanation**

 Here's a breakdown of the steps involved:

1. Data Preparation: The initial DataFrame `df` contains 4101 entries with 27 columns. The columns include features like 'Date' and various news articles labeled as 'News 1', 'News 2', ..., 'News 25', along with a target variable 'Label'. The 'Label' column is of integer type, while the news articles are of object type.

2. Feature Extraction: The code snippet `[22]` extracts the news articles from each row of the DataFrame and creates a list named 'news' containing the combined text of all the news articles for each row. The resulting 'news' list is a collection of text data.

3. Text Vectorization: The code snippet `[28]` imports the CountVectorizer from the scikit-learn library, which is used to convert the text data into numerical features. The CountVectorizer converts the text into a matrix of token counts. The `lowercase=True` parameter indicates that the text should be converted to lowercase, and `ngram_range=(1,1)` specifies that single words should be considered as features.

4. Data Transformation: The code snippet `[30]` applies the CountVectorizer to the 'news' list and transforms it into a numerical representation, stored in the variable 'x'. The resulting 'x' matrix has dimensions (4101, 48527), indicating 4101 rows (corresponding to the number of entries in the DataFrame) and 48527 columns (corresponding to the number of unique features extracted from the text data).

5. Target Variable and Train-Test Split: The code snippet `[32]` assigns the 'Label' column from the DataFrame to the variable 'y'. Then, the data is split into training and testing sets using the `train_test_split()` function from scikit-learn in `[36]`. The 'x' and 'y' variables are split into 'x_train', 'x_test', 'y_train', and 'y_test' using a test size of 30% and a stratified split based on the 'Label' variable.

6. Model Training: A Random Forest Classifier is initialized with 200 estimators in `[38]`, and it is fitted on the training data using `[39]`. The Random Forest Classifier is an ensemble learning method that combines multiple decision trees to make predictions.

7. Prediction and Evaluation: The trained model is used to predict the target variable for the test data in `[40]`, and the predictions are stored in 'y_pred'. The model's performance is evaluated using metrics such as confusion matrix, classification report, and accuracy score. The confusion matrix in `[42]` shows the number of true negatives, false positives, false negatives, and true positives. The classification report in `[43]` provides precision, recall, F1-score, and support for each class, along with macro and weighted averages.

In summary, the machine learning model uses a Random Forest Classifier to analyze the sentiment of financial market data based on news articles. The text data is transformed into numerical features using the CountVectorizer, and the model is trained and evaluated using a train-test split of the data. The model's performance can be assessed based on the provided evaluation metrics.